src/hotspot/share/opto/callnode.hpp

```diff
@@ -503,15 +503,76 @@ class SafePointNode : public MultiNode {
 };
 
 //------------------------------SafePointScalarObjectNode----------------------
-// A SafePointScalarObjectNode represents the state of a scalarized object
-// at a safepoint.
+// A SafePointScalarObjectNode (aka spobj) represents the state of one or more
+// scalarized objects at a SafePointNode. There are two scenarios where a
+// 'spobj' is created: 1) A scalar replaced (SR) object is directly referenced
+// by a SafePoint; 2) A scalar replaced object is participating in an allocation
+// merge (Phi) and the Phi is referenced by a SafePoint. The schematics of how
+// 'spobj' is used in both scenarios are described below.
+//
+// Scenario 1: Representing an object directly referenced by a SafePoint.
+//   Only the _first_index, _n_fields and _alloc fields are required.
+//   _first_index : relative index in the SafePoint node's input array where pointers
+//                  to the values of the fields of the SR object are going to be stored.
+//   _nfields     : how many fields the SR object has.
+//   _alloc       : pointer to the Allocate object that previously created the SR object.
+//                  Only used for debug purposes.
+//
+// Scenario 2: Representing SR objects participating in merges.
+//  Only the _merge_pointer_idx and _number_of_objects fields are required.
+//  _number_of_objects : how many objects participating in the merge were SR
+//                       and are represented in this 'sobj'.
+//  _merge_pointer_idx : relative index in the SafePoint node's input array
+//                       where the description of the SR _allocation merge_
+//                       starts. The two entries in the SafePoint node's input
+//                       array starting at '_merge_pointer_idx` are Phi nodes
+//                       representing: 1) a pointer to any non-SR'ed object
```

Comment:
For #1, if there was a single non-SR-ed object participating in the merge, is there still a Phi node referenced here? Or is it just a pointer to the non-SR'ed object?

Reply:
Scenario 1 doesn't involve Phi/merges. So, SafePointScalarObjectNode will describe a single SR'ed object. If you're curious, you can search macro.cpp for uses of SafePointScalarObjectNode.